Add OpInfo based meta tensor tests

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76905

https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Add OpInfo based meta tensor tests

https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

TODO: There are failures that correspond to known bugs and need to be
skipped.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

anjali411 commented on test/test_meta.py:
```diff
@@ -0,0 +1,837 @@
+                        suppress = True
+                if not suppress:
+                    raise RuntimeError(f"""\
+failed to run: {func}(
```
indentation

ezyang replied:
actually unindenting triple blocks like this is idiomatic, because indenting them will result in indentation in the underlying string

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

TODO: There are failures that correspond to known bugs and need to be
skipped.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

TODO: There are failures that correspond to known bugs and need to be
skipped.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9603,6 +9603,7 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
                # Reference: https://github.com/pytorch/pytorch/issues/50747
                DecorateInfo(unittest.skip("Skipped!"), 'TestCommon', 'test_variant_consistency_eager',
                             dtypes=all_types_and_complex_and(torch.bool, torch.bfloat16, torch.float16)),
+               DecorateInfo(unittest.expectedFailure, 'TestMeta', 'test_meta', dtypes=(torch.bool,)),
```
Nice xfails

ezyang commented:
@mruberry So a question: for stuff like this, I could have instead found a test where we're running forwards, and smooshed my test into that one. The tests are less isolated but they run faster (because instead of running the stock CPU/CUDA implementation twice, I instead only run it once).

mruberry commented on test/test_meta.py:
```diff
@@ -0,0 +1,846 @@
+
+RE_NOT_IMPLEMENTED_MSG = re.compile(r"Could not run '([^']+)' with arguments ")
+
+meta_exclude_set = {
```
holy exclude set, Batman!

mruberry commented on test/test_meta.py:
```diff
@@ -0,0 +1,846 @@
+    torch._validate_sparse_csr_tensor_args,
+}
+
+class MetaCrossRefMode(torch.overrides.TorchFunctionMode):
```
Comment pls

mruberry commented on test/test_meta.py:
```diff
@@ -0,0 +1,846 @@
+    @skipIfCrossRef
+    @suppress_warnings
+    @ops(op_db)
+    def test_meta(self, device, dtype, op):
```
Comment describing this test, too, would be nice

mruberry reviewed (approved):
Cool! A few inline suggestions but looks pretty good from a test framework perspective

ngimel commented on test/test_meta.py:
```diff
@@ -0,0 +1,846 @@
+    torch.Tensor.cholesky_solve,  # MISSING aten::_cholesky_solve_helper
+    torch.Tensor.clamp,  # MISSING aten::clamp.Tensor
+    torch.Tensor.clamp_,  # MISSING aten::clamp.Tensor_out
+    torch.Tensor.clamp_max,  # MISSING aten::clamp_max.out
```
Should be fixed by #76874 for scalar overload at least

ngimel replied:
And now tensor overloads are getting fixed too.

mruberry commented:
> @mruberry So a question: for stuff like this, I could have instead found a test where we're running forwards, and smooshed my test into that one. The tests are less isolated but they run faster (because instead of running the stock CPU/CUDA implementation twice, I instead only run it once).

Totally! This comes up a lot (our ops get called on the same sample inputs SO MANY TIMES!). And combining tests would make the test suite faster, but it would also impact test granularity. Like if the meta portion of a test failed then what about the dtypes part, or the autograd part, etc. What would be really interesting is if we could use subtest-like functionality to have these larger tests that preserved their computed artifacts for reuse across a variety of subtests, but we'd need to ensure the subtests could succeed or fail independently, and that we could still apply our xfail/skip/tolerance/etc. decorators to the subtests, too.

ezyang commented on test/test_meta.py:
```diff
@@ -0,0 +1,846 @@
+    @ops(op_db)
+    def test_meta(self, device, dtype, op):
+        func = op.get_op()
+        if func in meta_exclude_set:
```
I used to do this, but this makes the test suite run *a lot* longer, like 4x longer (because throwing exceptions is expensive). As a compromise, I can add an optional mode to do it this way, but not the default.

mruberry replied:
Yeah I understand it takes longer. If you don't want to do it let's not bother, we can add a mode later if we really want to

ezyang replied:
So, when I'm doing OpInfo tests only, it's tolerable. So I've turned on expected failures, and shuffled things around so that I can NOT do that for crossref testing.

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -201,6 +201,11 @@ Tensor& set_storage_cpu_(Tensor& result, Storage storage, int64_t storage_offset
   return result;
 }
 
+Tensor& set_(Tensor& result, const Tensor& storage, int64_t storage_offset, IntArrayRef size, IntArrayRef stride) {
```
nit: storage -> source

albanD commented on test/test_meta.py:
```diff
@@ -0,0 +1,900 @@
+# crossref testing, which means there is probably not enough coverage from
+# OpInfo.  Patch in https://github.com/pytorch/pytorch/pull/75994 and find
+# out where these fails come from.
+suspicious_exclude_set = {
```
nit: not used? do we want to keep it?

ezyang replied:
Yes I do, I'll use it for real crossref testing later

albanD reviewed (approved):
SGTM

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add OpInfo based meta tensor tests"


https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this

ezyang commented:
@pytorchbot revert this because it broke periodic tests
